New error behaviour, fixes #5

mrfishie opened the pull request:
Fixes https://github.com/IvanSanchez/node-artemis-lib/issues/5.
Implements two events for error behaviour: one for 'critical' errors such as bad magic number or bad header, the other for not-so-critical errors such as unknown type/subtype, length mismatch, etc.

mrfishie committed:
New error behaviour

ParseErrors include the buffer, and are split into two events: 'error'
for critical errors that close the connection, and 'warn' for
not-so-critical things (unknown packet type, length mismatch, etc)

mrfishie committed:
Replace warn with unparsed

See
https://github.com/IvanSanchez/node-artemis-lib/issues/5#issuecomment-95892467

mrfishie committed:
Prevents the socket from closing after a parse error is encountered

IvanSanchez commented on src/artemisSocket.js:
```diff
@@ -21,7 +17,14 @@ class Socket extends net.Socket {
-		this.on('data',this._parseData);
+		this.on('data', this._parseData);
+
+		this.on('error', this._onError);
```
I don't think this is needed - the functionality of `net.Socket` can take care of panicking on critical errors, I think.

mrfishie replied:
Thanks! Fixed in https://github.com/mrfishie/node-artemis-lib/commit/491cd23cb6eb7fd6c0086cad589deaf8f9673199

IvanSanchez commented on src/artemisSocket.js:
```diff
@@ -100,8 +100,7 @@ class Socket extends net.Socket {
 			} catch(err) {
-				this.emit('error', new ParseError('Could not parse packet (' + header.type.toString(16) + ',' + subtype.toString(16) + '): ' + err.message));
-				this._debugBuffer(buffer);
+				this.emit('error', new ParseError('Could not parse packet (' + header.type.toString(16) + ',' + subtype.toString(16) + '): ' + err.message, buffer));
```
`this.emit('unparsed');` !!!

mrfishie committed:
Removed unnecessary error event

net.Socket automatically closes when an error occurs (see
https://nodejs.org/api/net.html#net_event_error_1)

IvanSanchez commented:
:+1:
